### PR TITLE
* FEATURE : Provide `MemoryUsage` trace and emitted memory usage data.

### PR DESCRIPTION
**Pull request by shaopeng-gh**
Based on pattern of Michael's newest check-in, now it does not need interface changing and is straightforward.

### CODE REVIEW

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs [hunk 1]
```diff
@@ -1060,19 +1060,41 @@ public static void AnalyzeTargetHelper(TContext context, IEnumerable<Skimmer<TCo
...
                         string directory = Path.GetDirectoryName(file);
                         file = Path.GetFileName(file);
-                        string timing = $"'{file}' : elapsed {stopwatch.Elapsed} : '{skimmer.Name}' : at '{directory}'";
+                        string timing = $"'{file}' : elapsed {stopwatch.Elapsed} : '{skimmer.Id}.{skimmer.Name}' : at '{directory}'";
 
                         context.Logger.LogToolNotification(
                             new Notification
                             {
-                                Level = FailureLevel.Warning,
+                                Level = FailureLevel.Note,
```
Comment (shaopeng-gh) [inline src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:1097]: >Note [](http://example.com/codeflow?start=53&length=4)

The other overall time log is `Note`, suggest to use `Note` as well

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs [hunk 2]
```diff
@@ -1060,19 +1060,41 @@ public static void AnalyzeTargetHelper(TContext context, IEnumerable<Skimmer<TCo
...
                         : null;
 
+                    if (context.Traces.HasFlag(DefaultTraces.MemoryUsage))
+                    {
+                        using (var currentProcess = Process.GetCurrentProcess())
+                        {
+                            string file = context.TargetUri.LocalPath;
+                            string directory = Path.GetDirectoryName(file);
+                            file = Path.GetFileName(file);
+                            string memoryUsage = $"'{file}' : current process memory usage {currentProcess.WorkingSet64 / 1024 / 1024}MB : '{skimmer.Id}.{skimmer.Name}' : at '{directory}'. Peak memory usage: {currentProcess.PeakWorkingSet64 / 1024 / 1024}MB";
```
Comment (michaelcfanning) [inline src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:1070]: >memoryUsage [](http://example.com/codeflow?start=35&length=11)

Can you please reply here with a real world example of the new output?

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs [hunk 3]
```diff
@@ -1060,19 +1060,41 @@ public static void AnalyzeTargetHelper(TContext context, IEnumerable<Skimmer<TCo
                         ? Stopwatch.StartNew()
                         : null;
 
+                    if (context.Traces.HasFlag(DefaultTraces.MemoryUsage))
```
Comment (michaelcfanning) [inline src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:1063]: >context [](http://example.com/codeflow?start=24&length=7)

should we move this earlier in the method, before starting the stop watch? i.e., we don't need to include this code in the timing data.

Comment (shaopeng-gh) [reply src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:1065]: fixed

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs (hunk 2 above)
Comment (shaopeng-gh) [reply src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs:1072]: yes, I tried to use the same format as the one you added,

[this one:]
BINSKIM : note 'BinaryParsers.dll' : current process memory usage 39MB : 'BA2004.EnableSecureSourceCodeHashing' : at 'C:\GH\binskim\bld\bin\x64_Debug\netcoreapp3.1'. Peak memory usage: 39MB

[the one you added before:]
BINSKIM : note 'BinaryParsers.dll' : elapsed 00:00:00.0263046 : 'BA2004.EnableSecureSourceCodeHashing' : at 'C:\GH\binskim\bld\bin\x64_Debug\netcoreapp3.1'

File: src/Sarif/DefaultTraces.cs [hunk 4]
```diff
@@ -10,7 +10,8 @@ public enum DefaultTraces
     {
         None,
         ScanTime = 0x01,
-        ScanExecution = 0x2,
+        ScanExecution = 0x02,
         RuleScanTime = 0x04,
+        MemoryUsage = 0x08,
```
Comment (michaelcfanning) [inline src/Sarif/DefaultTraces.cs:15]: >MemoryUsage [](http://example.com/codeflow?start=8&length=11)

Should we have a global memory usage vs. peak mem data point for the end-to-end scan? And the one per rule/per target? Who is asking for this feature, by the way?

File: src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs (hunk 2 above)
Comment (shaopeng-gh) [reply src/Sarif.Driver/Sdk/MultithreadedAnalyzeCommandBase.cs]: Now that this is changed to log only once the new format is :
`BINSKIM : note Done. 1,386 files scanned, elapsed time 00:00:25.6855079.`
`BINSKIM : note Peak memory usage: 1213MB.`

File: src/Sarif/DefaultTraces.cs (hunk 4 above)
Comment (shaopeng-gh) [reply src/Sarif/DefaultTraces.cs:15]: fixed to only log once per end of the PR review.

**Pull request closed without merging**